src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -0,0 +1,200 @@
+
+bool MergeTreeBackgroundExecutor::trySchedule(ExecutableTaskPtr task)
+{
+    std::lock_guard lock(mutex);
+
+    if (shutdown)
+        return false;
+
+    try
+    {
+        /// This is needed to increase / decrease the number of threads at runtime
+        if (update_timer.compareAndRestartDeferred(10.))
+            updateConfiguration();
+    }
+    catch (...)
+    {
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+    }
+
+    auto & value = CurrentMetrics::values[metric];
+    if (value.load() >= static_cast<int64_t>(max_tasks_count))
+        return false;
+
+    pending.push_back(std::make_shared<Item>(std::move(task), metric));
+
+    has_tasks.notify_one();
+    return true;
+}
+
+
+void MergeTreeBackgroundExecutor::removeTasksCorrespondingToStorage(StorageID id)
+{
+    std::vector<ItemPtr> tasks_to_wait;
+    {
+        std::lock_guard lock(mutex);
+
+        /// Erase storage related tasks from pending and select active tasks to wait for
+        auto it = std::remove_if(pending.begin(), pending.end(),
+            [&] (auto item) -> bool { return item->task->getStorageID() == id; });
+        pending.erase(it, pending.end());
+
+        /// Copy items to wait for their completion
+        std::copy_if(active.begin(), active.end(), std::back_inserter(tasks_to_wait),
+            [&] (auto item) -> bool { return item->task->getStorageID() == id; });
+
+        for (auto & item : tasks_to_wait)
+            item->is_currently_deleting = true;
+    }
+
+
+    for (auto & item : tasks_to_wait)
+        item->is_done.wait();
+}
+
+
+void MergeTreeBackgroundExecutor::routine(ItemPtr item)
+{
+    auto erase_from_active = [this, item]
+    {
+        active.erase(std::remove(active.begin(), active.end(), item), active.end());
+    };
+
+    try
+    {
+        if (item->task->execute())
+        {
+            std::lock_guard guard(mutex);
+
+            if (item->is_currently_deleting)
+            {
+                erase_from_active();
+                return;
+            }
+
+            pending.push_back(item);
+            erase_from_active();
+            has_tasks.notify_one();
+            return;
+        }
+
+        std::lock_guard guard(mutex);
+        erase_from_active();
+        has_tasks.notify_one();
+        /// In a situation of a lack of memory this method can throw an exception,
+        /// because it may interact somehow with BackgroundSchedulePool, which may allocate memory
+        /// But it is rather safe, because we have try...catch block here, and another one in ThreadPool.
+        item->task->onCompleted();
+        item->task.reset();
+    }
+    catch (...)
+    {
+        std::lock_guard guard(mutex);
+        erase_from_active();
+        has_tasks.notify_one();
+        tryLogCurrentException(__PRETTY_FUNCTION__);
+        /// Do not want any exceptions
+        try { item->task->onCompleted(); } catch (...) {}
+    }
+}
+
+
+void MergeTreeBackgroundExecutor::threadFunction(size_t number)
```

Comment:
What is `number`?

Reply:
Added comment.

Reply:
Maybe we could also wrap the whole `threadFunction` into `try/catch`, and in case of exception
a) assert in debug
b) write very scary log message and run `threadFunction` again